### ISSUE TITLE
Stuck after starting Android emulator in locale en-US

### ISSUE THREAD

**Issue by mmcc007** [mmcc007/screenshots#29]
I'm seeing this issue as well with Android emulators (iOS simulators work fine), and am using only en-US:

screenshots.yaml:
```
...
locales:
#  - fr-CA
  - en-US # this is assumed to be the default for your test machine.
...
devices:
  ios:
    - iPhone X
    - iPad Pro (12.9-inch) (2nd generation)
  android:
    - Nexus 6P
...
```

Issue exists for both Nexus 5X and 6P.
```
$ emulator -list-avds
Nexus_5X_API_27
Nexus_6P_API_28
```

_Originally posted by @noumaans in https://github.com/mmcc007/screenshots/issues/19#issuecomment-477313724_

**Comment by mmcc007** [comment]
@noumaans Can you send me the output your are getting when running screenshots.
Should be something like:
https://travis-ci.com/mmcc007/screenshots/builds/104898262#L1054
(though, due to hardware acceleration, your emulator should be running much faster)

... and your 'flutter doctor -v'

BTW: does your app use internationalization?

thnx.

**Comment by noumaans** [comment]
Output of screenshots:
```
$ screenshots
Clearing images in android/fastlane/metadata/android/en-US/images/phoneScreenshots for 'Nexus 6P'...
Starting emulator 'Nexus_6P' in locale en-US ...
```
I can see the emulator launched and it appears to be ready, but screenshots seems to be waiting for emulator.

This is happening for the default flutter app, created simply by:
flutter create myapp

```
$ flutter doctor -v
[✓] Flutter (Channel beta, v1.3.8, on Mac OS X 10.14.3 18D109, locale en-US)
    • Flutter version 1.3.8 at /Users/noumaan/flutter
    • Framework revision e5b1ed7a7f (4 weeks ago), 2019-03-06 14:23:37 -0800
    • Engine revision f4951df193
    • Dart version 2.2.1 (build 2.2.1-dev.0.0 571ea80e11)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/noumaan/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /Users/noumaan/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[!] iOS toolchain - develop for iOS devices (Xcode 10.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2, Build version 10E125
    ✗ Verify that all connected devices have been paired with this computer in Xcode.
      If all devices have been paired, libimobiledevice and ideviceinstaller may require updating.
      To update with Brew, run:
        brew update
        brew uninstall --ignore-dependencies libimobiledevice
        brew uninstall --ignore-dependencies usbmuxd
        brew install --HEAD usbmuxd
        brew unlink usbmuxd
        brew link usbmuxd
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    • ios-deploy 1.9.2
    ✗ ios-deploy out of date (1.9.4 is required). To upgrade with Brew:
        brew upgrade ios-deploy
    • CocoaPods version 1.5.2

[✓] Android Studio (version 3.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 33.4.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.32.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.24.0

[!] Connected device
    ! No devices available
```

**Comment by noumaans** [comment]
Please update title of this issue to:
"Stuck after starting Android emulator in locale en-US"

**Comment by mmcc007** [comment]
It looks like it is not finding a (exact) matching emulator for 'Nexus 6P' as mentioned in README.
The validator is also not catching it.
The choice is to either fix the validator, or add a feature to provide a more flexible criteria for finding a matching emulator. Both should resolve this issue.

Since the feature is already implemented for iOS, I figured it makes sense to implement it for android also.

You can try the fix using:
```
git clone https://github.com/mmcc007/screenshots.git
cd screenshots
git checkout issue_29
pub global activate --source path .
```

If you can confirm that this resolves the problem on your machine, then I will considered it fixed.

**Comment by Pitazzo** [comment]
I've the same issue here, both with Nexus 6P and and Nexus 5X; the emulator launches but screenshots keep waiting for it to start. I've installed the `issue_29 `git branch and the only difference was this output: 
```
$ screenshots -c screenshots.yaml 
Warning: Exact match for 'Nexus 6P' not found.
       : Using 'Nexus_6P_API_28'.
Clearing images in android/fastlane/metadata/android/es-ES/images/phoneScreenshots for 'Nexus 6P'...
Starting emulator 'Nexus 6P' in locale es-ES ...
```

**Comment by mmcc007** [comment]
@Pitazzo I added a test for this condition and ran it in:
https://travis-ci.com/mmcc007/screenshots/builds/107011036
It is finding the emulator and starting it successfully. 

There could possibly be an installation issue on your machine. Can you check if 'adb is finding the emulator by running:
```
adb devices
or
flutter devices
```
while _screenshots_ is waiting for the emulator?

You could also try something like:
```
/tmp/screenshots/resources/script/android-wait-for-emulator
```

Also, can you send me your 'flutter doctor -v'?

Thnx.

**Comment by Pitazzo** [comment]
@mmcc007 thanks for your quick reply. I followed your instructions with these results.

While screenshots was waiting, I run `flutter devices` and the emulator was found: 
```
1 connected device:

Android SDK built for x86 • emulator-5554 • android-x86 • Android 9 (API 28) (emulator)
```
Then, when I executed `adb devices`, I found out that `abd` wasn't installed in my system. After installing it, I was able to run `adb kill-server`. Now, screenshots was finally able to launch my app and I managed to see the its splash screen. I don't know if `abd` should be considered as a dependency of screenshots.

However, the app didn't completely launch and I got this exception in screenshots:
```
Warning: 'Nexus 6P' does not have a matching emulator.
       : Using 'Nexus_6P_API_28'.
Clearing images in android/fastlane/metadata/android/es-ES/images/phoneScreenshots for 'Nexus 6P'...
Starting emulator 'Nexus 6P' in locale es-ES ...
Emulator is ready
Capturing screenshots with test test/screenshots_tests.dart on emulator Nexus 6P in locale es-ES ...
Using device Android SDK built for x86.
Starting application: test/screenshots_tests.dart
Initializing gradle...                                              0,8s
Resolving dependencies...                                           2,6s
Installing build/app/outputs/apk/app.apk...                         1,3s
Running Gradle task 'assembleDebug'...                              5,1s
Built build/app/outputs/apk/debug/app-debug.apk.
I/flutter ( 5030): Observatory listening on http://127.0.0.1:34803/
Stopping application instance.
test_driver/screenshots_tests_test.dart:6:8: Error: Not found: 'dart:ui'
import 'dart:ui';
       ^
file:///home/pitazzo/SDKs/flutter/packages/flutter_test/lib/src/accessibility.dart:8:8: Error: Not found: 'dart:ui'
import 'dart:ui' as ui;
       ^
file:///home/pitazzo/SDKs/flutter/packages/flutter_test/lib/src/binding.dart:8:8: Error: Not found: 'dart:ui'
import 'dart:ui' as ui;
       ^
file:///home/pitazzo/SDKs/flutter/packages/flutter_test/lib/src/matchers.dart:8:8: Error: Not found: 'dart:ui'
import 'dart:ui' as ui;
       ^
file:///home/pitazzo/SDKs/flutter/packages/flutter_test/lib/src/matchers.dart:9:8: Error: Not found: 'dart:ui'
import 'dart:ui';
       ^
file:///home/pitazzo/SDKs/flutter/packages/flutter_test/lib/src/test_pointer.dart:12:1: Error: Not found: 'dart:ui'
export 'dart:ui' show Offset;
^
file:///home/pitazzo/SDKs/flutter/packages/flutter_test/lib/src/window.dart:6:8: Error: Not found: 'dart:ui'
import 'dart:ui' hide window;
       ^
file:///home/pitazzo/SDKs/flutter/packages/flutter/lib/src/material/animated_icons.dart:9:8: Error: Not found: 'dart:ui'
import 'dart:ui' as ui show Paint, Path, Canvas;
       ^
file:///home/pitazzo/SDKs/flutter/packages/flutter/lib/src/material/animated_icons.dart:10:8: Error: Not found: 'dart:ui'
import 'dart:ui' show lerpDouble;
       ^
file:///home/pitazzo/SDKs/flutter/packages/flutter/lib/src/material/app.dart:5:8: Error: Not found: 'dart:ui'
import 'dart:ui' as ui;
       ^
Driver tests failed: 254
Unhandled exception:
command failed: cmd='flutter drive test/screenshots_tests.dart'
#0      streamCmd (package:screenshots/utils.dart:66:7)
<asynchronous suspension>
#1      screenshots (package:screenshots/screenshots.dart:102:9)
<asynchronous suspension>
#2      run (package:screenshots/screenshots.dart:53:17)
<asynchronous suspension>
#3      main (file:///home/pitazzo/Proyectos/screenshots/bin/main.dart:75:9)
<asynchronous suspension>
#4      _startIsolate.<anonymous closure> (dart:isolate/runtime/libisolate_patch.dart:298:32)
#5      _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)
```

This is my `flutter doctor -v` output:
```
[✓] Flutter (Channel stable, v1.2.1, on Linux, locale es_ES.UTF-8)
    • Flutter version 1.2.1 at /home/pitazzo/SDKs/flutter
    • Framework revision 8661d8aecd (hace 7 semanas), 2019-02-14 19:19:53 -0800
    • Engine revision 3757390fa4
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /home/pitazzo/SDKs/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /home/pitazzo/SDKs/Android/Sdk
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[!] Android Studio (version 3.3)
    • Android Studio at /opt/android-studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] VS Code (version 1.32.3)
    • VS Code at /usr/share/code
    • Flutter extension version 2.25.0

[✓] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 9 (API 28) (emulator)

! Doctor found issues in 1 category.
```

One last thing: screenshots expected to find my test file in the folder test_driver instead of in the path I specified in the .yaml conif file; I don't know it that's the correct behaviour. I had to move there my test file in order to get the error I described.

Thanks for your time

**Comment by mmcc007** [comment]
Glad you were able to get _screenshots_ to get the emulator started for you!

'flutter devices' uses some clever coding to find 'adb'. Usually, no matter which way you install Android SDK, 'adb' is included in the install. But 'adb' is often not included in the PATH. 

_screenshots_ expects to be able to find 'adb' in the PATH. So that is probably why it was hanging.
I added a check for 'adb' at the start to highlight this.

The 'dart:ui' runtime error is probably not related to _screenshots_. The following might help:
https://github.com/flutter/flutter/issues/27826

'flutter driver' generally expects the name of the app (which has the 'enableFlutterDriverExtension()') and the related test to be in the same directory. In practice this means creating a test app that calls your app and creating the related test. This pair of files should be in the same directory (usually test_driver) and not in the 'lib' directory. This is also not related to _screenshots_ as you should be getting the same error from running 'flutter driver' from the command line.

Can you confirm that this issue #29 has been resolved for you?

**Comment by noumaans** [comment]
> @noumaans It looks like it is not finding a (exact) matching emulator for 'Nexus 6P' as mentioned in README.
> The validator is also not catching it.
> The choice is to either fix the validator, or add a feature to provide a more flexible criteria for finding a matching emulator. Both should resolve this issue.
> 
> Since the feature is already implemented for iOS, I figured it makes sense to implement it for android also.
> 
> You can try the fix using:
> 
> ```
> git clone https://github.com/mmcc007/screenshots.git
> cd screenshots
> git checkout issue_29
> pub global activate --source path .
> ```
> If you can confirm that this resolves the problem on your machine, then I will considered it fixed.


Thanks for the fix. It's working now. Here's the output at my end:
```
$ flutter packages pub global run screenshots:main
Warning: 'Nexus 6P' does not have a matching emulator.
       : Using 'Nexus_6P_API_28'.
Clearing images in android/fastlane/metadata/android/en-US/images/phoneScreenshots for 'Nexus 6P'...
Starting emulator 'Nexus 6P' in locale en-US ...
Waiting for emulator to start: 599 : status: error: device offline
Waiting for emulator to start: 598 : status: error: device offline
Waiting for emulator to start: 597 : status: error: device offline
Waiting for emulator to start: 596 : status: error: device offline
Waiting for emulator to start: 595 : status: error: device offline
Waiting for emulator to start: 594 : status: error: device offline
Waiting for emulator to start: 593 : status: error: device offline
Waiting for emulator to start: 592 : status: error: device offline
Emulator is ready
Capturing screenshots with test test_driver/app.dart on emulator Nexus 6P in locale en-US ...
Using device Android SDK built for x86.
Starting application: test_driver/app.dart
Initializing gradle...                                              5.9s
Resolving dependencies...                                          11.0s
Installing build/app/outputs/apk/app.apk...                        13.4s
Running Gradle task 'assembleDebug'...                             14.8s
Built build/app/outputs/apk/debug/app-debug.apk.
Installing build/app/outputs/apk/app.apk...                         5.4s
I/flutter ( 6099): Observatory listening on http://127.0.0.1:33651/
00:00 +0: Fastlane App (setUpAll)
00:02 +0: Fastlane App starts at 0
Screenshot counter_0 created
00:06 +1: Fastlane App increments the counter
Screenshot counter_1 created
00:14 +2: Fastlane App (tearDownAll)
00:14 +2: All tests passed!
Stopping application instance.
[info ] FlutterDriver: Connecting to Flutter application at http://127.0.0.1:56001/
[trace] FlutterDriver: Isolate found with number: 691227379
[trace] FlutterDriver: Isolate is paused at start.
[trace] FlutterDriver: Attempting to resume isolate
[trace] FlutterDriver: Waiting for service extension
[info ] FlutterDriver: Connected to Flutter application.
Processing screenshots from test...
Moving screenshots to android/fastlane/metadata/android/en-US/images/phoneScreenshots
Stopping emulator: Nexus_6P_API_28 ...
OK: killing emulator, bye bye
OK
Emulator is stopped


Screen images are available in:
  ios/fastlane/screenshots
  android/fastlane/metadata/android
```

**Comment by mmcc007** [comment]
Great @noumaans!

I consider this issue resolved and have merged into master and included in latest release (v0.1.7).

Thanx @noumaans and @Pitazzo for the feedback!
